domain/storage/service/storage.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/collections/transform"
+
+	coreerrors "github.com/juju/juju/core/errors"
+	"github.com/juju/juju/core/logger"
+	corestorage "github.com/juju/juju/core/storage"
+	"github.com/juju/juju/domain/storage"
+	storageerrors "github.com/juju/juju/domain/storage/errors"
+	"github.com/juju/juju/environs/envcontext"
+	"github.com/juju/juju/environs/tags"
+	"github.com/juju/juju/internal/errors"
+	internalstorage "github.com/juju/juju/internal/storage"
+)
+
+// StorageState defines an interface for interacting with the underlying state.
+type StorageState interface {
+	// GetModelDetails returns the model and controller UUID for the current model.
+	GetModelDetails() (storage.ModelDetails, error)
+	// ImportFilesystem associates a filesystem (either native or volume backed) hosted by a cloud provider
+	// with a new storage instance (and storage pool) in a model.
+	ImportFilesystem(ctx context.Context, name internalstorage.Name,
+		filesystem storage.FilesystemInfo) (internalstorage.ID, error)
+}
+
+// StorageService defines a service for storage related behaviour.
+type StorageService struct {
+	st             State
+	logger         logger.Logger
+	registryGetter corestorage.ModelStorageRegistryGetter
+}
+
+// ImportFilesystem associates a filesystem (either native or volume backed) hosted by a cloud provider
+// with a new storage instance (and storage pool) in a model.
+// The following error types can be expected:
+// - [coreerrors.NotSupported]: when the importing the kind of storage is not supported by the provider.
+// - [storageerrors.InvalidPoolNameError]: when the supplied pool name is invalid.
+func (s *StorageService) ImportFilesystem(
+	ctx context.Context, credentialInvalidatorGetter envcontext.ModelCredentialInvalidatorGetter, arg ImportStorageParams,
+) (internalstorage.ID, error) {
+	if arg.Kind != internalstorage.StorageKindFilesystem {
+		// TODO(axw) implement support for volumes.
```

Comment:
I think we've done this right?

Reply:
The CLI command is `import-filesystem`. We do support importing volume backed filesystems. There's never been though fully implemented and tested support for importing block device storage. Hence the todo.